libscript/sources/Interpreter.cxx

```diff
@@ -0,0 +1,531 @@
+#include "Logger.hpp"
+#include "pivot/script/Builtins.hxx"
+#include "pivot/script/Exceptions.hxx"
+#include <limits>
+#include <unordered_map>
+
+namespace pivot::ecs::script::interpreter
+{
+
+// TODO: consider using magicenum
+const std::map<NodeType, std::string> gNodeTypeStrings = {
+    {NodeType::File, "File"},
+    {NodeType::ComponentDeclaration, "ComponentDeclaration"},
+    {NodeType::SystemDeclaration, "SystemDeclaration"},
+    {NodeType::ComponentName, "ComponentName"},
+    {NodeType::SystemName, "SystemName"},
+    {NodeType::PropertyType, "PropertyType"},
+    {NodeType::PropertyName, "PropertyName"},
+    {NodeType::EventKeyword, "EventKeyword"},
+    {NodeType::EventName, "EventName"},
+    {NodeType::EventPayloadType, "EventPayloadType"},
+    {NodeType::EventPayloadName, "EventPayloadName"},
+    {NodeType::EntityParameterName, "EntityParameterName"},
+    {NodeType::EntityParameterComponent, "EntityParameterComponent"},
+    {NodeType::Symbol, "Symbol"}};
+const std::map<std::string, data::BasicType> gVariableTypes{{"Vector3", data::BasicType::Vec3},
+                                                            {"Number", data::BasicType::Number},
+                                                            {"Boolean", data::BasicType::Boolean},
+                                                            {"Color", data::BasicType::Number},
+                                                            {"String", data::BasicType::String}};
+// Map builtin binary (two operands) operators, to their callback
+const std::map<std::string, std::function<data::Value(const data::Value &, const data::Value &)>> gOperatorCallbacks = {
+    {"*", interpreter::builtins::builtin_operatorMul},   {"/", interpreter::builtins::builtin_operatorDiv},
+    {"%", interpreter::builtins::builtin_operatorMod},   {"+", interpreter::builtins::builtin_operatorAdd},
+    {"-", interpreter::builtins::builtin_operatorSub},   {"<", interpreter::builtins::builtin_operatorLt},
+    {"<=", interpreter::builtins::builtin_operatorLtEq}, {">", interpreter::builtins::builtin_operatorGt},
+    {">=", interpreter::builtins::builtin_operatorGtEq}, {"==", interpreter::builtins::builtin_operatorEq},
+    {"!=", interpreter::builtins::builtin_operatorNEq},  {"&&", interpreter::builtins::builtin_operatorAnd},
+    {"||", interpreter::builtins::builtin_operatorOr}};
+// Map built-in function strings, to their callback and parameters (number and types)
+const std::unordered_map<              // map
+    std::string, std::pair<            // function string name, to a pair
+                     std::function<    // first : callback
+                         data::Value(const std::vector<data::Value> &)>,
+                     std::pair<size_t, std::vector<std::vector<data::Type>>>    // second: pair; first: number of
+                                                                                // params, second: types of params
+                     >>
```

Comment:
shesh, that is a complicated type

consider using something like this for clarity
```suggestion
/// doc
using FunctionCallback = std::function<data::Value(const std::vector<data::Value> &)>
/// Doc
using ParameterPair = std::pair<size_t, std::vector<std::vector<data::Type>>>

const std::unordered_map<              // map
    std::string, std::pair<FunctionCallback, ParameterPair>>
```